python/cudf/cudf/core/indexing_utils.py

```diff
@@ -0,0 +1,225 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import enum
+import itertools
+from typing import TYPE_CHECKING, Any, List, Tuple, Union
+
+import numpy as np
+from typing_extensions import TypeAlias
+
+import cudf
+import cudf._lib as libcudf
+from cudf.api.types import (
+    _is_scalar_or_zero_d_array,
+    is_bool_dtype,
+    is_integer,
+    is_integer_dtype,
+)
+
+if TYPE_CHECKING:
+    from cudf.core.column import ColumnBase
+
+
+class IndexTag(enum.IntEnum):
+    SLICE = enum.auto()
+    MASK = enum.auto()
+    MAP = enum.auto()
+    SCALAR = enum.auto()
+
+
+IndexSpec: TypeAlias = Tuple[IndexTag, Union[slice, "ColumnBase"]]
+ColumnLabels: TypeAlias = List[str]
+
+
+def unpack_iloc_key(
+    key: Any, frame: cudf.DataFrame | cudf.Series, n: int
+) -> List[Any]:
+    """Unpack a user-level key to iloc.__getitem__
```

Comment:
What does it mean to "unpack"? Even just a quick example as part of the description here would greatly help readers of this function.

Reply:
Ditto below for unpack_dataframe_iloc_indexer

Reply:
I think this is perhaps now a better implementation:
```python


def destructure_iloc_key(
    key: Any, frame: cudf.Series | cudf.DataFrame
) -> tuple[Any, ...]:
    """
    Destructure a potentially tuple-typed key into row and column indexers

    Tuple arguments to iloc indexing are treated specially. They are
    picked apart into indexers for the row and column. If the number
    of entries is less than the number of modes of the frame, missing
    entries are slice-expanded.

    If the user-provided key is not a tuple, it is treated as if it
    were a singleton tuple, and then slice-expanded.

    Once this destructuring has occurred, any entries that are
    callables are then called with the indexed frame. This should
    return a valid indexing object for the rows (respectively
    columns), namely one of:

    - A boolean mask of the same length as the frame in the given
      dimension
    - A scalar integer that indexes the frame
    - An array-like of integers that index the frame
    - A slice that indexes the frame

    Integer and slice-based indexing follows usual Python conventions.

    Parameters
    ----------
    key
        The key to destructure
    frame
        DataFrame or Series to provide context

    Returns
    -------
    tuple of length equal to the dimension of the frame of indexers

    Raises
    ------
    IndexError
        If there are too many indexers, or any individual indexer is a tuple.
    """
    n = len(frame)
    if isinstance(key, tuple):
        indexers = tuple(
            itertools.chain(key, itertools.repeat(slice(None), n - len(key)))
        )
        if (ni := len(indexers)) > n:
            raise IndexError(f"Too many indexers: got {ni} expected {n}")
    else:
        indexers = (key, *itertools.repeat(slice(None), n - 1))
    indexers = tuple(k(frame) if callable(k) else k for k in indexers)
    if any(isinstance(k, tuple) for k in indexers):
        raise IndexError(
            "Too many indexers: can't have nested tuples in iloc indexing"
        )
    return indexers
```

Reply:
> Ditto below for unpack_dataframe_iloc_indexer

This is doing a little bit more. It destructures into row and column indexers (using the pattern above), and then transforms the column indexer into column names.

Reply:
Used that suggestion.